auth/src/krb.cpp

```diff
@@ -0,0 +1,942 @@
+        gss_name_t myName = GSS_C_NO_NAME;
+        gss_name_t myName2 = GSS_C_NO_NAME;
+
+        // generate a gss name struct for the specified name
+        if (!_specified_name.empty()) {
+            gss_buffer_desc name_buf{};
+            name_buf.value = strdup(_specified_name.c_str());
+            name_buf.length = _specified_name.length() + 1;
+            majorStatus = gss_import_name(&minorStatus,
+                                          &name_buf,
+                                          const_cast<gss_OID>(gss_nt_service_name_krb),
+                                          &myName);
+            if (majorStatus != GSS_S_COMPLETE) {
+                const std::string msg = fmt::format(fmt::runtime(
+                                                    "Failed importing specified name: \"{}\"."),
+                                                    _specified_name);
+
+                krb_log_error(_r_error, msg.data(), majorStatus, minorStatus, _is_client);
+
+                THROW(KRB_ERROR_IMPORT_NAME, msg);
+            }
+        }
+
+        // Acquire the credentials if we do not have any yet
+        if (my_creds == GSS_C_NO_CREDENTIAL) {
+            majorStatus = gss_acquire_cred(
+                    &minorStatus,
+                    myName,
+                    0,
+                    GSS_C_NULL_OID_SET,
+                    _specified_name.empty() ? GSS_C_INITIATE : GSS_C_ACCEPT,
+                    &my_creds,
+                    NULL,
+                    NULL);
+        }
+        else {
+            // Already have credentials so all is well
+            majorStatus = GSS_S_COMPLETE;
+        }
+
+        if (majorStatus != GSS_S_COMPLETE) {
+            static const char* msg = "Failed acquiring credentials.";
+            krb_log_error(_r_error, msg, majorStatus, minorStatus, _is_client);
+            THROW(KRB_ERROR_ACQUIRING_CREDS, msg);
+        }
+
+        // set the credentials in the auth object
+        gss_release_name(&minorStatus, &myName);
+
+        majorStatus = gss_inquire_cred(&minorStatus, my_creds, &myName2, NULL, NULL, NULL);
+        if (majorStatus != GSS_S_COMPLETE) {
+            static const char* msg = "Failed inquiring creds for the name.";
+            krb_log_error(_r_error, msg, majorStatus, minorStatus, _is_client);
+            THROW(KRB_ERROR_ACQUIRING_CREDS, msg);
+        }
+
+        gss_OID doid2;
+        gss_buffer_desc client_name2;
+        majorStatus = gss_display_name(&minorStatus, myName2, &client_name2, &doid2);
+        if (majorStatus != GSS_S_COMPLETE) {
+            static const char* msg = "Failed during displaying name.";
+            krb_log_error(_r_error, msg, majorStatus, minorStatus, _is_client);
+            THROW(KRB_ERROR_DISPLAYING_NAME, msg);
+        }
+
+        std::string display_name;
+        if (client_name2.value && client_name2.length > 0) {
+            display_name = std::string{static_cast<char*>(client_name2.value)};
```

Comment:
Shouldn't need the `std::string{...}` wrapping, but not a blocker.

Reply:
Will fix

Reply:
Done